src/Reserve.sol

```diff
@@ -4,30 +4,115 @@ pragma solidity ^0.8.7;
 import {Ownable} from "openzeppelin-contracts/access/Ownable.sol";
 import {IERC20} from "openzeppelin-contracts/token/ERC20/IERC20.sol";
 
+/// @notice The reserve interface as seen by the users.
 interface IReserve {
-    function withdraw(
+    /// @notice Deposits funds into the reserve.
+    /// The reserve will `transferFrom` `amt` tokens from the `from` address.
+    /// @param token The used token.
+    /// @param from The address from which funds are deposited.
+    /// @param amt The deposited amount.
+    function deposit(
         IERC20 token,
-        address to,
+        address from,
         uint256 amt
     ) external;
 
-    function deposit(
+    /// @notice Withdraws funds from the reserve.
+    /// The reserve will transfer `amt` tokens to the `to` address.
+    /// Only funds previously deposited can be withdrawn.
+    /// @param token The used token.
+    /// @param to The address to which funds are withdrawn.
+    /// @param amt The withdrawn amount.
+    function withdraw(
         IERC20 token,
-        address from,
+        address to,
         uint256 amt
     ) external;
 }
 
+/// @notice The reserve plugin interface required by the reserve.
+interface IReservePlugin {
+    /// @notice Called by the reserve immediately after transferring funds for deposition.
+    /// @param token The used token.
+    /// @param amt The amount which has been transferred for deposition.
+    function afterDeposition(IERC20 token, uint256 amt) external;
+
+    /// @notice Called by the reserve right before transferring funds for withdrawal.
+    /// The reserve will `transferFrom` the tokens from the plugin address.
+    /// The reserve can always withdraw everything that has been ever deposited, but never more.
+    /// @param token The used token.
+    /// @param amt The amount which will be transferred.
+    function beforeWithdrawal(IERC20 token, uint256 amt) external;
+}
+
+/// @notice The ERC-20 tokens reserve contract.
+/// The registered users can deposit and withdraw funds.
+/// The reserve by default doesn't do anything with the tokens,
+/// but for each ERC-20 address a plugin can be registered for tokens storage.
 contract Reserve is IReserve, Ownable {
+    /// @notice The dummy plugin address meaning that no plugin is being used.
+    IReservePlugin public constant NO_PLUGIN = IReservePlugin(address(0));
+
+    /// @notice A set of addresses considered users.
+    /// The value is `true` if an address is a user, `false` otherwise.
     mapping(address => bool) public isUser;
```

Comment:
I think i would prefer `access`/ `allowance` here because the we use the term `user` already in the DripsHub context and ReserveUsers are other Drips smart contracts and not the Drips user

Reply:
However, no strong opinion.

Reply:
I too am not a fan of `user`. Maybe `controller`?

Reply:
Maybe `Depositor` ?

Reply:
Haha. I depends how you see it. You are seeing the entire system.  I see it only from the perspective of the Reserve. There is no DripsHub from the perspective of the Reserve.

Just smart contracts or EOA which can deposit funds.  Now, we need a good name for those addresses.

Here is my name canditate:
`depositor` =>  `a person who keeps money at a bank`. 
(See: https://dictionary.cambridge.org/de/worterbuch/englisch/depositor)

Which might be a good name candidate for your reserve user.